### PR TITLE
TC-13: Add new endpoint to calculate tax amount for all the states.

### PR DESCRIPTION
**Pull request by creddy-cpal**

### CODE REVIEW

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java [hunk 1]
```diff
@@ -1,29 +1,63 @@
...
+        taxPrice.setStateCode(typeData.getCode());
+        taxPrice.setTotalAmount(price.add(taxableAmount));
+        return taxPrice;
+    }
+
+    private BigDecimal getStateTaxRate(String state) {
+        return stateDAO.findStateTaxRate(state);
+    }
+
+    private BigDecimal calculateTaxableAmount(BigDecimal price, BigDecimal taxRate) {
```
Comment (sbandaru-cpal) [inline services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java]: Can you create a TaxUtil class common/common-utils module

File: services/tax-calculator/tax-calculator-api/src/main/java/com/cpal/tax/calculator/api/TaxPrice.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package com.cpal.tax.calculator.api;
+
+import java.math.BigDecimal;
+
+public class TaxPrice {
+
+    private BigDecimal taxRate;
+    private String stateCode;
```
Comment (sbandaru-cpal) [inline services/tax-calculator/tax-calculator-api/src/main/java/com/cpal/tax/calculator/api/TaxPrice.java]: this needs to be State enum

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/dao/StateDAO.java [hunk 3]
```diff
@@ -21,8 +21,8 @@
         return STATES;
     }
     
-    public BigDecimal findStateTaxRate(State state) {
-        return StateTax.valueOf(state.name()).getTaxRate();
+    public BigDecimal findStateTaxRate(String state) {
```
Comment (sbandaru-cpal) [inline services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/dao/StateDAO.java]: keep it as State enum, need to restrict to only the states in the State enum

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java [hunk 4]
```diff
@@ -1,29 +1,63 @@
...
+
+    public BigDecimal calculateTaxedTotalPrice(State state, BigDecimal price) {
+        BigDecimal taxPercent = getStateTaxRate(state.name());
+        BigDecimal taxableAmount = calculateTaxableAmount(price, taxPercent);
+        return price.add(taxableAmount);
+
+    }
+
+    public List<TaxPrice> calculateTotalPriceForAllStates(BigDecimal price) {
+        List<TypeData> states = getAllStates();
```
Comment (sbandaru-cpal) [inline services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java]: loop through State.values()

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/controller/TaxController.java [hunk 5]
```diff
@@ -31,4 +32,9 @@
     public @ResponseBody BigDecimal calculate(@PathVariable State state, @PathVariable BigDecimal price) {
         return stateService.calculateTaxedTotalPrice(state, price);
     }
+
+    @RequestMapping(method = RequestMethod.GET, value = "/calculateAllStates/{price}")
```
Comment (sbandaru-cpal) [inline services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/controller/TaxController.java]: calculate/allStates/{price}

**Review by sbandaru-cpal — CHANGES_REQUESTED** [review]
Address the comments

File: services/tax-calculator/tax-calculator-api/src/main/java/com/cpal/tax/calculator/api/TaxPrice.java (hunk 2 above)
Comment (creddy-cpal) [reply services/tax-calculator/tax-calculator-api/src/main/java/com/cpal/tax/calculator/api/TaxPrice.java]: Fixed..

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/controller/TaxController.java (hunk 5 above)
Comment (creddy-cpal) [reply services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/controller/TaxController.java]: Fixed

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/dao/StateDAO.java (hunk 3 above)
Comment (creddy-cpal) [reply services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/dao/StateDAO.java]: Fixed..

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java (hunk 4 above)
Comment (creddy-cpal) [reply services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java]: Fixed.

File: services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java (hunk 1 above)
Comment (creddy-cpal) [reply services/tax-calculator/tax-calculator-service/src/main/java/com/cpal/tax/calculator/service/StateService.java]: Fixed.

**Review by sbandaru-cpal — APPROVED** [review]

**Pull request merged**